### ISSUE TITLE
Trying to send signed email even though TLS is disabled

### ISSUE THREAD

**Issue by adngdb** [wekan/wekan#1120]
Hello,

In my admin panel, I have disabled TLS for sending emails. However, when trying to recover a password for example, the client shows "Internal server error" and the logs show this: 

```
wekan_1    | Exception while invoking method 'forgotPassword' Error: self signed certificate
wekan_1    |     at Object.Future.wait (/build/programs/server/node_modules/fibers/future.js:449:15)
wekan_1    |     at Mail._syncSendMail (packages/meteor.js:213:24)
wekan_1    |     at smtpSend (packages/email.js:110:13)
wekan_1    |     at Object.Email.send (packages/email.js:168:5)
wekan_1    |     at AccountsServer.Accounts.sendResetPasswordEmail (packages/accounts-password/password_server.js:614:9)
wekan_1    |     at [object Object].Meteor.methods.forgotPassword (packages/accounts-password/password_server.js:546:12)
wekan_1    |     at packages/check.js:130:16
wekan_1    |     at [object Object]._.extend.withValue (packages/meteor.js:1122:17)
wekan_1    |     at Object.exports.Match._failIfArgumentsAreNotAllChecked (packages/check.js:129:41)
wekan_1    |     at maybeAuditArgumentChecks (packages/ddp-server/livedata_server.js:1734:18)
wekan_1    |     at packages/ddp-server/livedata_server.js:719:19
wekan_1    |     at [object Object]._.extend.withValue (packages/meteor.js:1122:17)
wekan_1    |     at packages/ddp-server/livedata_server.js:717:40
wekan_1    |     at [object Object]._.extend.withValue (packages/meteor.js:1122:17)
wekan_1    |     at packages/ddp-server/livedata_server.js:715:46
wekan_1    |     at [object Object]._.extend.protocol_handlers.method (packages/ddp-server/livedata_server.js:689:23)
wekan_1    |     - - - - -
wekan_1    |     at Error (native)
wekan_1    |     at TLSSocket.<anonymous> (_tls_wrap.js:1063:38)
wekan_1    |     at emitNone (events.js:67:13)
wekan_1    |     at TLSSocket.emit (events.js:166:7)
wekan_1    |     at TLSSocket._init.ssl.onclienthello.ssl.oncertcb.TLSSocket._finishInit (_tls_wrap.js:621:8)
wekan_1    |     at TLSWrap.ssl.onclienthello.ssl.oncertcb.ssl.onnewsession.ssl.onhandshakedone (_tls_wrap.js:453:38)
```

Why is wekan trying to sign anything even though TLS is disabled? Is it a bug or am I missing something? 

Instance is installed with docker-compose, image is the latest one at the time of writing (0.27), and my email server config has been confirmed working with a number of other services. 

Thanks!

**Comment by xet7** [comment]
@adngdb 

Do you have same correct settings for MAIL_URL in docker-compose.yml and Admin Panel ?

Does sending email work without TLS with some other program on server?

It's complaining about self-signed certificate. You could add paid or free Let's Encrypt TLS certificate to your email SMTP server.

Also see:
https://github.com/wekan/wekan/wiki/Troubleshooting-Mail

**Comment by Nate-in-AK** [comment]
We're seeing the same issue on a snap based install.  Mail works fine from command line.

**Comment by anlx-sw** [comment]
same issue. for me - also a snap installation. it worked for some time before but with one of the recent updates the mail connector seems to use tls although defined as 'smtp://internalmailrelay:25/' using plain smtp without auth.

**Comment by xet7** [comment]
Does this still happen on Wekan v0.28 that includes SMTP fix #1131 ?

**Comment by anlx-sw** [comment]
yes wekan-snap is already running on version 0.28 and the error is still reproducable.
![grafik](https://user-images.githubusercontent.com/16399949/28330021-4562d67c-6bec-11e7-89a0-7e8d58ac9c94.png)

here is the full error ouput:

```
Jul 18 19:09:12 srvkanban01 snap[1858]: Exception while invoking method 'forgotPassword' Error: unable to get local issuer certificate
Jul 18 19:09:12 srvkanban01 snap[1858]:     at Object.Future.wait (/snap/wekan-ondra/101/programs/server/node_modules/fibers/future.js:449:15)
Jul 18 19:09:12 srvkanban01 snap[1858]:     at Mail._syncSendMail (packages/meteor.js:213:24)
Jul 18 19:09:12 srvkanban01 snap[1858]:     at smtpSend (packages/email.js:116:13)
Jul 18 19:09:12 srvkanban01 snap[1858]:     at Object.Email.send (packages/email.js:174:5)
Jul 18 19:09:12 srvkanban01 snap[1858]:     at AccountsServer.Accounts.sendResetPasswordEmail (packages/accounts-password/password_server.js:614:9)
Jul 18 19:09:12 srvkanban01 snap[1858]:     at [object Object].Meteor.methods.forgotPassword (packages/accounts-password/password_server.js:546:12)
Jul 18 19:09:12 srvkanban01 snap[1858]:     at packages/check.js:130:16
Jul 18 19:09:12 srvkanban01 snap[1858]:     at [object Object]._.extend.withValue (packages/meteor.js:1122:17)
Jul 18 19:09:12 srvkanban01 snap[1858]:     at Object.exports.Match._failIfArgumentsAreNotAllChecked (packages/check.js:129:41)
Jul 18 19:09:12 srvkanban01 snap[1858]:     at maybeAuditArgumentChecks (packages/ddp-server/livedata_server.js:1734:18)
Jul 18 19:09:12 srvkanban01 snap[1858]:     at packages/ddp-server/livedata_server.js:719:19
Jul 18 19:09:12 srvkanban01 snap[1858]:     at [object Object]._.extend.withValue (packages/meteor.js:1122:17)
Jul 18 19:09:12 srvkanban01 snap[1858]:     at packages/ddp-server/livedata_server.js:717:40
Jul 18 19:09:12 srvkanban01 snap[1858]:     at [object Object]._.extend.withValue (packages/meteor.js:1122:17)
Jul 18 19:09:12 srvkanban01 snap[1858]:     at packages/ddp-server/livedata_server.js:715:46
Jul 18 19:09:12 srvkanban01 snap[1858]:     at [object Object]._.extend.protocol_handlers.method (packages/ddp-server/livedata_server.js:689:23)
Jul 18 19:09:12 srvkanban01 snap[1858]:     - - - - -
Jul 18 19:09:12 srvkanban01 snap[1858]:     at Error (native)
Jul 18 19:09:12 srvkanban01 snap[1858]:     at TLSSocket.<anonymous> (_tls_wrap.js:1065:38)
Jul 18 19:09:12 srvkanban01 snap[1858]:     at emitNone (events.js:67:13)
Jul 18 19:09:12 srvkanban01 snap[1858]:     at TLSSocket.emit (events.js:166:7)
Jul 18 19:09:12 srvkanban01 snap[1858]:     at TLSSocket._init.ssl.onclienthello.ssl.oncertcb.TLSSocket._finishInit (_tls_wrap.js:623:8)
Jul 18 19:09:12 srvkanban01 snap[1858]:     at TLSWrap.ssl.onclienthello.ssl.oncertcb.ssl.onnewsession.ssl.onhandshakedone (_tls_wrap.js:453:38)



```

**Comment by mpartipilo** [comment]
I **FINALLY** figured out the problem... It's the docker-versatile-postfix that includes a self-signed certificate in its image. I bash'd into the container and commented these lines in /etc/postfix/main.cf and now email works:
```
# TLS parameters
#smtpd_tls_cert_file=/etc/ssl/certs/ssl-cert-snakeoil.pem
#smtpd_tls_key_file=/etc/ssl/private/ssl-cert-snakeoil.key
#smtpd_use_tls=yes
#smtpd_tls_session_cache_database = btree:${data_directory}/smtpd_scache
#smtp_tls_session_cache_database = btree:${data_directory}/smtp_scache
```

I will start looking for a replacement for that docker image. Perhaps go with https://github.com/catatnight/docker-postfix or https://github.com/tomav/docker-mailserver

**Comment by anlx-sw** [comment]
@mpartipilo - just to check  if i understand you right::
there is a error with wekan on  docker when a locally installed smtpd in the docker image is used and offering tls? (so you configure wekan to use smtp://127.0.0.1:25/  - without tls and it is still connecting with tls?)

if this is the case -  there still remains the issue that wekan tries to send with tls even if it is told to use cleartext smtp. 
i run into issues using the snap implementation (and not docker). 

or is my understanding wrong?
should i open a seperate issue?

**Comment by mpartipilo** [comment]
You understand right.

Config is not specifying TLS at all, connection goes to port 25 using
smtp://user:pass@emailserver:25/

The config in the admin panel is also saved without TLS checked.

Disabling TLS on the SMTP server works. I guess the problem there stems
from a self signed cert inside the postfix container.

The smtpd is not running locally with wekan though. It runs in a separate
container.

This issue began around the update to Meteor 1.4, somewhere around v0.23.

I don't know snap, but I bet it's the same problem.

**Comment by anlx-sw** [comment]
i just found something in the meteor email pagkage implementation. it seems they added STARTTLS support for the email package: https://github.com/meteor/meteor/issues/8637.

theory:
so in my case i try to send to a mail relay which was contacted without a STARTTLS header before. Now with the recent package it does STARTTLS. The problem seems to be that the certificate from the mail relay presented when using  STARTTLS is untrusted/unknown in the linux system (or maybe it has to be trusted within the snap). i try to search if this theory is right and i can get it working by adding the certificates. i couldn't find a way to disable STARTTLS for the meteor email package.

**Comment by anlx-sw** [comment]
ok - so my ubuntu server had the certificates already in the store and a 
`openssl s_client -starttls smtp -crlf -connect internalmailrelay::25`
showed that everything should be fine.

unfortunately if i switch the shell into the contect of the snap via
`snap run --shell wekan-ondra.help`
i cant use openssl there to test if everything is working - snap is blocking this system calls.

on futher search i found out that snap itself changed the behavior and is no longer using the certificate auth store now but an internal one for snap: https://github.com/snapcore/snapd/pull/3271

if found no possiblity to configure this internal store so far - but I'll try to dig deeper in the rabbit hole tomorrow ;-)

**Comment by xet7** [comment]
@kubiko 

Do you have ideas how to fix this?

**Comment by kubiko** [comment]
@anlx-sw for your testing, use 
`snap run --shell wekan.database-backup`
as help command does not have network plugs, therefore apparmor would flag calls to /etc/resolvconf
I tested with database-backup and 
`openssl s_client -starttls smtp -crlf -connect smtp.fastmail.com:465`
works fine for me
BTW @anlx-sw please switch from wekan-ondra to wekan snap, as I'm planning to phase it out

**Comment by kubiko** [comment]
@anlx-sw have you tested trick they proposed on meteor?
`MAIL_URL='smtp://localhost:1025' -> MAIL_URL='smtp://user:pw123@localhost:1025'`
this seems to trick the metor to not use tls...

**Comment by anlx-sw** [comment]
@kubiko - thanks for your help. 
with the network enabled shell i could verify that indeed this is the issue. the snap can't verify the certificate - although the server can:
snap
` verify error:num=20:unable to get local issuer certificate`
server
`verify return:1`

snapd is only using the ca-certifictes under 
` /snap/core/current/etc/ssl/certs`
this directories are read only - so you can't add anything to it.

setting a user and password didn't help for me 
I set 
`MAIL_URL='smtp://foo:bar@internalmailrelay:25/'`
and restarted wekan. on the next test mail i still get the same error.

i can't find any currently supported way to add certificates to the snap but i found a discussion concerning this issue:
https://forum.snapcraft.io/t/extending-system-certificates/114/12

so basically there is currently no option to disable ssl as meteor is always using STARTTLS if the server is supporting it.
and there is no supported option to configure snapd in a way that is uses the sytem cert store - as it was the case with older builds.

so you are stuck to public ca's already in the ca store of snap currently it seems.

has anyone an idea for a clever workarround?

**Comment by anlx-sw** [comment]
as a dirty workaround we configured dedicated smtp connectors for the wekan server which are not offering STARTTLS. this solved the problem for my installation.

i think this problem will still bite every snap user with an internal mail relay which is offering starttls or enforcing tls. and it is especially nasty to troubleshoot if you configure smtp without TLS and the meteor email provider is using it anyways. 
so a better solution to the problem should be found.

**Comment by kubiko** [comment]
not having that much time, but I had quick look and I can add openssl into the snap package, then we can configure it properly. As once it's built it can't be configured, and for obvious reason one in core snap is looking at /etc/ssl/certs
If we add openssl inside snap, then we can configure it to look to
$SNAP_COMMON/etc/ssl/certs
If I manage to make test build happy to test it for me?

**Comment by xet7** [comment]
I did today some STARTTLS debugging, and got it working with following, you can test:
https://github.com/wekan/wekan/wiki/Troubleshooting-Mail#starttls-sending-test-email-failed

**Comment by xet7** [comment]
https://github.com/wekan/wekan/wiki/Troubleshooting-Mail#example-aws-ses
